{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dna = 'CTGTTGACATGCATTCACGCTACGCTAGCT'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dna=dna.replace('C', 'M')\n",
      "dna=dna.replace('G', 'C')\n",
      "dna=dna.replace('M', 'G')\n",
      "dna=dna.replace('A', 'M')\n",
      "dna=dna.replace('T', 'A')\n",
      "dna=dna.replace('M', 'T')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print(dna)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Functions that always produce the same value aren't particularly useful, so most functions take input values called parameters or arguments. A parameter is a variable that is assigned a value each time the function is called. For example, this function that converts a temperature from Fahrenheit to Kelvin has one parameter called temp:"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def <function name>():\n",
      "    <function body>"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def <function name>(<argument>):\n",
      "    <function body>\n",
      "    return <local variable 1> "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dna = 'CTGTTGACATGCATTCACGCTACGCTAGCT'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def dna_sequence_complement(dna):\n",
      "    dna=dna.replace('C', 'M')\n",
      "    dna=dna.replace('G', 'C')\n",
      "    dna=dna.replace('M', 'G')\n",
      "    dna=dna.replace('A', 'M')\n",
      "    dna=dna.replace('T', 'A')\n",
      "    dna=dna.replace('M', 'T')\n",
      "    return dna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print dna_sequence_complement(dna)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "GACAACTGTACGTAAGTGCGATGCGATCGA\n"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "neandertal = 'ACTTCCGCTTCCTCAACAGCAGCACTTCCGGGTTGGGAGAAAGGTGG'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print dna_sequence_complement(neandertal)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "TGAAGGCGAAGGAGTTGTCGTCGTGAAGGCCCAACCCTCTTTCCACC\n"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def dna_complement(sequence):\n",
      "    sequence=sequence.replace('C', 'M')\n",
      "    sequence=sequence.replace('G', 'C')\n",
      "    sequence=sequence.replace('M', 'G')\n",
      "    sequence=sequence.replace('A', 'M')\n",
      "    sequence=sequence.replace('T', 'A')\n",
      "    sequence=sequence.replace('M', 'T')\n",
      "    return sequence"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print dna_complement(neandertal)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "TGAAGGCGAAGGAGTTGTCGTCGTGAAGGCCCAACCCTCTTTCCACC\n"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "neandertal = 'ACTTCCGCTTCCTCAACAGCAGCACTTCCGGGTTGGGAGAAAGGTGG'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 42
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dna_complement(neandertal)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 43,
       "text": [
        "'TGAAGGCGAAGGAGTTGTCGTCGTGAAGGCCCAACCCTCTTTCCACC'"
       ]
      }
     ],
     "prompt_number": 43
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Seq: A read-only sequence object (essentially a string with an alphabet)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio.Seq import Seq\n",
      "\n",
      "my_dna = Seq(\"AGTACACTGGT\")\n",
      "\n",
      "Seq.complement(my_dna)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 28,
       "text": [
        "Seq('TCATGTGACCA', Alphabet())"
       ]
      }
     ],
     "prompt_number": 28
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "my_dna.reverse_complement()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 30,
       "text": [
        "Seq('ACCAGTGTACT', Alphabet())"
       ]
      }
     ],
     "prompt_number": 30
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "dir(Seq)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 29,
       "text": [
        "['__add__',\n",
        " '__class__',\n",
        " '__cmp__',\n",
        " '__contains__',\n",
        " '__delattr__',\n",
        " '__dict__',\n",
        " '__doc__',\n",
        " '__format__',\n",
        " '__getattribute__',\n",
        " '__getitem__',\n",
        " '__hash__',\n",
        " '__init__',\n",
        " '__len__',\n",
        " '__module__',\n",
        " '__new__',\n",
        " '__radd__',\n",
        " '__reduce__',\n",
        " '__reduce_ex__',\n",
        " '__repr__',\n",
        " '__setattr__',\n",
        " '__sizeof__',\n",
        " '__str__',\n",
        " '__subclasshook__',\n",
        " '__weakref__',\n",
        " '_get_seq_str_and_check_alphabet',\n",
        " 'back_transcribe',\n",
        " 'complement',\n",
        " 'count',\n",
        " 'endswith',\n",
        " 'find',\n",
        " 'lower',\n",
        " 'lstrip',\n",
        " 'reverse_complement',\n",
        " 'rfind',\n",
        " 'rsplit',\n",
        " 'rstrip',\n",
        " 'split',\n",
        " 'startswith',\n",
        " 'strip',\n",
        " 'tomutable',\n",
        " 'tostring',\n",
        " 'transcribe',\n",
        " 'translate',\n",
        " 'ungap',\n",
        " 'upper']"
       ]
      }
     ],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "help(Seq)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Help on class Seq in module Bio.Seq:\n",
        "\n",
        "class Seq(__builtin__.object)\n",
        " |  A read-only sequence object (essentially a string with an alphabet).\n",
        " |  \n",
        " |  Like normal python strings, our basic sequence object is immutable.\n",
        " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
        " |  Seq objects to be used as dictionary keys.\n",
        " |  \n",
        " |  The Seq object provides a number of string like methods (such as count,\n",
        " |  find, split and strip), which are alphabet aware where appropriate.\n",
        " |  \n",
        " |  In addition to the string like sequence, the Seq object has an alphabet\n",
        " |  property. This is an instance of an Alphabet class from Bio.Alphabet,\n",
        " |  for example generic DNA, or IUPAC DNA. This describes the type of molecule\n",
        " |  (e.g. RNA, DNA, protein) and may also indicate the expected symbols\n",
        " |  (letters).\n",
        " |  \n",
        " |  The Seq object also provides some biological methods, such as complement,\n",
        " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
        " |  not applicable to sequences with a protein alphabet).\n",
        " |  \n",
        " |  Methods defined here:\n",
        " |  \n",
        " |  __add__(self, other)\n",
        " |      Add another sequence or string to this sequence.\n",
        " |      \n",
        " |      If adding a string to a Seq, the alphabet is preserved:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import generic_protein\n",
        " |      >>> Seq(\"MELKI\", generic_protein) + \"LV\"\n",
        " |      Seq('MELKILV', ProteinAlphabet())\n",
        " |      \n",
        " |      When adding two Seq (like) objects, the alphabets are important.\n",
        " |      Consider this example:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet.IUPAC import unambiguous_dna, ambiguous_dna\n",
        " |      >>> unamb_dna_seq = Seq(\"ACGT\", unambiguous_dna)\n",
        " |      >>> ambig_dna_seq = Seq(\"ACRGT\", ambiguous_dna)\n",
        " |      >>> unamb_dna_seq\n",
        " |      Seq('ACGT', IUPACUnambiguousDNA())\n",
        " |      >>> ambig_dna_seq\n",
        " |      Seq('ACRGT', IUPACAmbiguousDNA())\n",
        " |      \n",
        " |      If we add the ambiguous and unambiguous IUPAC DNA alphabets, we get\n",
        " |      the more general ambiguous IUPAC DNA alphabet:\n",
        " |      \n",
        " |      >>> unamb_dna_seq + ambig_dna_seq\n",
        " |      Seq('ACGTACRGT', IUPACAmbiguousDNA())\n",
        " |      \n",
        " |      However, if the default generic alphabet is included, the result is\n",
        " |      a generic alphabet:\n",
        " |      \n",
        " |      >>> Seq(\"\") + ambig_dna_seq\n",
        " |      Seq('ACRGT', Alphabet())\n",
        " |      \n",
        " |      You can't add RNA and DNA sequences:\n",
        " |      \n",
        " |      >>> from Bio.Alphabet import generic_dna, generic_rna\n",
        " |      >>> Seq(\"ACGT\", generic_dna) + Seq(\"ACGU\", generic_rna)\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      TypeError: Incompatible alphabets DNAAlphabet() and RNAAlphabet()\n",
        " |      \n",
        " |      You can't add nucleotide and protein sequences:\n",
        " |      \n",
        " |      >>> from Bio.Alphabet import generic_dna, generic_protein\n",
        " |      >>> Seq(\"ACGT\", generic_dna) + Seq(\"MELKI\", generic_protein)\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      TypeError: Incompatible alphabets DNAAlphabet() and ProteinAlphabet()\n",
        " |  \n",
        " |  __cmp__(self, other)\n",
        " |      Compare the sequence to another sequence or a string (README).\n",
        " |      \n",
        " |      Historically comparing Seq objects has done Python object comparison.\n",
        " |      After considerable discussion (keeping in mind constraints of the\n",
        " |      Python language, hashes and dictionary support) a future release of\n",
        " |      Biopython will change this to use simple string comparison. The plan is\n",
        " |      that comparing incompatible alphabets (e.g. DNA to RNA) will trigger a\n",
        " |      warning.\n",
        " |      \n",
        " |      This version of Biopython still does Python object comparison, but with\n",
        " |      a warning about this future change. During this transition period,\n",
        " |      please just do explicit comparisons:\n",
        " |      \n",
        " |      >>> seq1 = Seq(\"ACGT\")\n",
        " |      >>> seq2 = Seq(\"ACGT\")\n",
        " |      >>> id(seq1) == id(seq2)\n",
        " |      False\n",
        " |      >>> str(seq1) == str(seq2)\n",
        " |      True\n",
        " |      \n",
        " |      Note - This method indirectly supports ==, < , etc.\n",
        " |  \n",
        " |  __contains__(self, char)\n",
        " |      Implements the 'in' keyword, like a python string.\n",
        " |      \n",
        " |      e.g.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import generic_dna, generic_rna, generic_protein\n",
        " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\", generic_dna)\n",
        " |      >>> \"AAA\" in my_dna\n",
        " |      True\n",
        " |      >>> Seq(\"AAA\") in my_dna\n",
        " |      True\n",
        " |      >>> Seq(\"AAA\", generic_dna) in my_dna\n",
        " |      True\n",
        " |      \n",
        " |      Like other Seq methods, this will raise a type error if another Seq\n",
        " |      (or Seq like) object with an incompatible alphabet is used:\n",
        " |      \n",
        " |      >>> Seq(\"AAA\", generic_rna) in my_dna\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      TypeError: Incompatable alphabets DNAAlphabet() and RNAAlphabet()\n",
        " |      >>> Seq(\"AAA\", generic_protein) in my_dna\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      TypeError: Incompatable alphabets DNAAlphabet() and ProteinAlphabet()\n",
        " |  \n",
        " |  __getitem__(self, index)\n",
        " |      Returns a subsequence of single letter, use my_seq[index].\n",
        " |  \n",
        " |  __hash__(self)\n",
        " |      Hash for comparison.\n",
        " |      \n",
        " |      See the __cmp__ documentation - we plan to change this!\n",
        " |  \n",
        " |  __init__(self, data, alphabet=Alphabet())\n",
        " |      Create a Seq object.\n",
        " |      \n",
        " |      Arguments:\n",
        " |       - seq      - Sequence, required (string)\n",
        " |       - alphabet - Optional argument, an Alphabet object from Bio.Alphabet\n",
        " |      \n",
        " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
        " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
        " |      the seq property.\n",
        " |      \n",
        " |      However, will often want to create your own Seq objects directly:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
        " |      ...              IUPAC.protein)\n",
        " |      >>> my_seq\n",
        " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
        " |      >>> print my_seq\n",
        " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
        " |      >>> my_seq.alphabet\n",
        " |      IUPACProtein()\n",
        " |  \n",
        " |  __len__(self)\n",
        " |      Returns the length of the sequence, use len(my_seq).\n",
        " |  \n",
        " |  __radd__(self, other)\n",
        " |      Adding a sequence on the left.\n",
        " |      \n",
        " |      If adding a string to a Seq, the alphabet is preserved:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import generic_protein\n",
        " |      >>> \"LV\" + Seq(\"MELKI\", generic_protein)\n",
        " |      Seq('LVMELKI', ProteinAlphabet())\n",
        " |      \n",
        " |      Adding two Seq (like) objects is handled via the __add__ method.\n",
        " |  \n",
        " |  __repr__(self)\n",
        " |      Returns a (truncated) representation of the sequence for debugging.\n",
        " |  \n",
        " |  __str__(self)\n",
        " |      Returns the full sequence as a python string, use str(my_seq).\n",
        " |      \n",
        " |      Note that Biopython 1.44 and earlier would give a truncated\n",
        " |      version of repr(my_seq) for str(my_seq).  If you are writing code\n",
        " |      which need to be backwards compatible with old Biopython, you\n",
        " |      should continue to use my_seq.tostring() rather than str(my_seq).\n",
        " |  \n",
        " |  back_transcribe(self)\n",
        " |      Returns the DNA sequence from an RNA sequence. New Seq object.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\",\n",
        " |      ...                     IUPAC.unambiguous_rna)\n",
        " |      >>> messenger_rna\n",
        " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG', IUPACUnambiguousRNA())\n",
        " |      >>> messenger_rna.back_transcribe()\n",
        " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())\n",
        " |      \n",
        " |      Trying to back-transcribe a protein or DNA sequence raises an\n",
        " |      exception:\n",
        " |      \n",
        " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
        " |      >>> my_protein.back_transcribe()\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      ValueError: Proteins cannot be back transcribed!\n",
        " |  \n",
        " |  complement(self)\n",
        " |      Returns the complement sequence. New Seq object.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> my_dna = Seq(\"CCCCCGATAG\", IUPAC.unambiguous_dna)\n",
        " |      >>> my_dna\n",
        " |      Seq('CCCCCGATAG', IUPACUnambiguousDNA())\n",
        " |      >>> my_dna.complement()\n",
        " |      Seq('GGGGGCTATC', IUPACUnambiguousDNA())\n",
        " |      \n",
        " |      You can of course used mixed case sequences,\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import generic_dna\n",
        " |      >>> my_dna = Seq(\"CCCCCgatA-GD\", generic_dna)\n",
        " |      >>> my_dna\n",
        " |      Seq('CCCCCgatA-GD', DNAAlphabet())\n",
        " |      >>> my_dna.complement()\n",
        " |      Seq('GGGGGctaT-CH', DNAAlphabet())\n",
        " |      \n",
        " |      Note in the above example, ambiguous character D denotes\n",
        " |      G, A or T so its complement is H (for C, T or A).\n",
        " |      \n",
        " |      Trying to complement a protein sequence raises an exception.\n",
        " |      \n",
        " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
        " |      >>> my_protein.complement()\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      ValueError: Proteins do not have complements!\n",
        " |  \n",
        " |  count(self, sub, start=0, end=2147483647)\n",
        " |      Non-overlapping count method, like that of a python string.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name,\n",
        " |      which does a non-overlapping count!\n",
        " |      \n",
        " |      Returns an integer, the number of occurrences of substring\n",
        " |      argument sub in the (sub)sequence given by [start:end].\n",
        " |      Optional arguments start and end are interpreted as in slice\n",
        " |      notation.\n",
        " |      \n",
        " |      Arguments:\n",
        " |       - sub - a string or another Seq object to look for\n",
        " |       - start - optional integer, slice start\n",
        " |       - end - optional integer, slice end\n",
        " |      \n",
        " |      e.g.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_seq = Seq(\"AAAATGA\")\n",
        " |      >>> print my_seq.count(\"A\")\n",
        " |      5\n",
        " |      >>> print my_seq.count(\"ATG\")\n",
        " |      1\n",
        " |      >>> print my_seq.count(Seq(\"AT\"))\n",
        " |      1\n",
        " |      >>> print my_seq.count(\"AT\", 2, -1)\n",
        " |      1\n",
        " |      \n",
        " |      HOWEVER, please note because python strings and Seq objects (and\n",
        " |      MutableSeq objects) do a non-overlapping search, this may not give\n",
        " |      the answer you expect:\n",
        " |      \n",
        " |      >>> \"AAAA\".count(\"AA\")\n",
        " |      2\n",
        " |      >>> print Seq(\"AAAA\").count(\"AA\")\n",
        " |      2\n",
        " |      \n",
        " |      A non-overlapping search would give the answer as three!\n",
        " |  \n",
        " |  endswith(self, suffix, start=0, end=2147483647)\n",
        " |      Does the Seq end with the given suffix?  Returns True/False.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Return True if the sequence ends with the specified suffix\n",
        " |      (a string or another Seq object), False otherwise.\n",
        " |      With optional start, test sequence beginning at that position.\n",
        " |      With optional end, stop comparing sequence at that position.\n",
        " |      suffix can also be a tuple of strings to try.  e.g.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
        " |      >>> my_rna.endswith(\"UUG\")\n",
        " |      True\n",
        " |      >>> my_rna.endswith(\"AUG\")\n",
        " |      False\n",
        " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
        " |      True\n",
        " |      >>> my_rna.endswith((\"UCC\",\"UCA\",\"UUG\"))\n",
        " |      True\n",
        " |  \n",
        " |  find(self, sub, start=0, end=2147483647)\n",
        " |      Find method, like that of a python string.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Returns an integer, the index of the first occurrence of substring\n",
        " |      argument sub in the (sub)sequence given by [start:end].\n",
        " |      \n",
        " |      Arguments:\n",
        " |       - sub - a string or another Seq object to look for\n",
        " |       - start - optional integer, slice start\n",
        " |       - end - optional integer, slice end\n",
        " |      \n",
        " |      Returns -1 if the subsequence is NOT found.\n",
        " |      \n",
        " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
        " |      >>> my_rna.find(\"AUG\")\n",
        " |      3\n",
        " |  \n",
        " |  lower(self)\n",
        " |      Returns a lower case copy of the sequence.\n",
        " |      \n",
        " |      This will adjust the alphabet if required. Note that the IUPAC alphabets\n",
        " |      are upper case only, and thus a generic alphabet must be substituted.\n",
        " |      \n",
        " |      >>> from Bio.Alphabet import Gapped, generic_dna\n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_seq = Seq(\"CGGTACGCTTATGTCACGTAG*AAAAAA\", Gapped(IUPAC.unambiguous_dna, \"*\"))\n",
        " |      >>> my_seq\n",
        " |      Seq('CGGTACGCTTATGTCACGTAG*AAAAAA', Gapped(IUPACUnambiguousDNA(), '*'))\n",
        " |      >>> my_seq.lower()\n",
        " |      Seq('cggtacgcttatgtcacgtag*aaaaaa', Gapped(DNAAlphabet(), '*'))\n",
        " |      \n",
        " |      See also the upper method.\n",
        " |  \n",
        " |  lstrip(self, chars=None)\n",
        " |      Returns a new Seq object with leading (left) end stripped.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Optional argument chars defines which characters to remove.  If\n",
        " |      omitted or None (default) then as for the python string method,\n",
        " |      this defaults to removing any white space.\n",
        " |      \n",
        " |      e.g. print my_seq.lstrip(\"-\")\n",
        " |      \n",
        " |      See also the strip and rstrip methods.\n",
        " |  \n",
        " |  reverse_complement(self)\n",
        " |      Returns the reverse complement sequence. New Seq object.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> my_dna = Seq(\"CCCCCGATAGNR\", IUPAC.ambiguous_dna)\n",
        " |      >>> my_dna\n",
        " |      Seq('CCCCCGATAGNR', IUPACAmbiguousDNA())\n",
        " |      >>> my_dna.reverse_complement()\n",
        " |      Seq('YNCTATCGGGGG', IUPACAmbiguousDNA())\n",
        " |      \n",
        " |      Note in the above example, since R = G or A, its complement\n",
        " |      is Y (which denotes C or T).\n",
        " |      \n",
        " |      You can of course used mixed case sequences,\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import generic_dna\n",
        " |      >>> my_dna = Seq(\"CCCCCgatA-G\", generic_dna)\n",
        " |      >>> my_dna\n",
        " |      Seq('CCCCCgatA-G', DNAAlphabet())\n",
        " |      >>> my_dna.reverse_complement()\n",
        " |      Seq('C-TatcGGGGG', DNAAlphabet())\n",
        " |      \n",
        " |      Trying to complement a protein sequence raises an exception:\n",
        " |      \n",
        " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
        " |      >>> my_protein.reverse_complement()\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      ValueError: Proteins do not have complements!\n",
        " |  \n",
        " |  rfind(self, sub, start=0, end=2147483647)\n",
        " |      Find from right method, like that of a python string.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Returns an integer, the index of the last (right most) occurrence of\n",
        " |      substring argument sub in the (sub)sequence given by [start:end].\n",
        " |      \n",
        " |      Arguments:\n",
        " |       - sub - a string or another Seq object to look for\n",
        " |       - start - optional integer, slice start\n",
        " |       - end - optional integer, slice end\n",
        " |      \n",
        " |      Returns -1 if the subsequence is NOT found.\n",
        " |      \n",
        " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
        " |      >>> my_rna.rfind(\"AUG\")\n",
        " |      15\n",
        " |  \n",
        " |  rsplit(self, sep=None, maxsplit=-1)\n",
        " |      Right split method, like that of a python string.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Return a list of the 'words' in the string (as Seq objects),\n",
        " |      using sep as the delimiter string.  If maxsplit is given, at\n",
        " |      most maxsplit splits are done COUNTING FROM THE RIGHT.\n",
        " |      If maxsplit is omitted, all splits are made.\n",
        " |      \n",
        " |      Following the python string method, sep will by default be any\n",
        " |      white space (tabs, spaces, newlines) but this is unlikely to\n",
        " |      apply to biological sequences.\n",
        " |      \n",
        " |      e.g. print my_seq.rsplit(\"*\",1)\n",
        " |      \n",
        " |      See also the split method.\n",
        " |  \n",
        " |  rstrip(self, chars=None)\n",
        " |      Returns a new Seq object with trailing (right) end stripped.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Optional argument chars defines which characters to remove.  If\n",
        " |      omitted or None (default) then as for the python string method,\n",
        " |      this defaults to removing any white space.\n",
        " |      \n",
        " |      e.g. Removing a nucleotide sequence's polyadenylation (poly-A tail):\n",
        " |      \n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_seq = Seq(\"CGGTACGCTTATGTCACGTAGAAAAAA\", IUPAC.unambiguous_dna)\n",
        " |      >>> my_seq\n",
        " |      Seq('CGGTACGCTTATGTCACGTAGAAAAAA', IUPACUnambiguousDNA())\n",
        " |      >>> my_seq.rstrip(\"A\")\n",
        " |      Seq('CGGTACGCTTATGTCACGTAG', IUPACUnambiguousDNA())\n",
        " |      \n",
        " |      See also the strip and lstrip methods.\n",
        " |  \n",
        " |  split(self, sep=None, maxsplit=-1)\n",
        " |      Split method, like that of a python string.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Return a list of the 'words' in the string (as Seq objects),\n",
        " |      using sep as the delimiter string.  If maxsplit is given, at\n",
        " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
        " |      splits are made.\n",
        " |      \n",
        " |      Following the python string method, sep will by default be any\n",
        " |      white space (tabs, spaces, newlines) but this is unlikely to\n",
        " |      apply to biological sequences.\n",
        " |      \n",
        " |      e.g.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
        " |      >>> my_aa = my_rna.translate()\n",
        " |      >>> my_aa\n",
        " |      Seq('VMAIVMGR*KGAR*L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
        " |      >>> my_aa.split(\"*\")\n",
        " |      [Seq('VMAIVMGR', HasStopCodon(ExtendedIUPACProtein(), '*')), Seq('KGAR', HasStopCodon(ExtendedIUPACProtein(), '*')), Seq('L', HasStopCodon(ExtendedIUPACProtein(), '*'))]\n",
        " |      >>> my_aa.split(\"*\",1)\n",
        " |      [Seq('VMAIVMGR', HasStopCodon(ExtendedIUPACProtein(), '*')), Seq('KGAR*L', HasStopCodon(ExtendedIUPACProtein(), '*'))]\n",
        " |      \n",
        " |      See also the rsplit method:\n",
        " |      \n",
        " |      >>> my_aa.rsplit(\"*\",1)\n",
        " |      [Seq('VMAIVMGR*KGAR', HasStopCodon(ExtendedIUPACProtein(), '*')), Seq('L', HasStopCodon(ExtendedIUPACProtein(), '*'))]\n",
        " |  \n",
        " |  startswith(self, prefix, start=0, end=2147483647)\n",
        " |      Does the Seq start with the given prefix?  Returns True/False.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Return True if the sequence starts with the specified prefix\n",
        " |      (a string or another Seq object), False otherwise.\n",
        " |      With optional start, test sequence beginning at that position.\n",
        " |      With optional end, stop comparing sequence at that position.\n",
        " |      prefix can also be a tuple of strings to try.  e.g.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
        " |      >>> my_rna.startswith(\"GUC\")\n",
        " |      True\n",
        " |      >>> my_rna.startswith(\"AUG\")\n",
        " |      False\n",
        " |      >>> my_rna.startswith(\"AUG\", 3)\n",
        " |      True\n",
        " |      >>> my_rna.startswith((\"UCC\",\"UCA\",\"UCG\"),1)\n",
        " |      True\n",
        " |  \n",
        " |  strip(self, chars=None)\n",
        " |      Returns a new Seq object with leading and trailing ends stripped.\n",
        " |      \n",
        " |      This behaves like the python string method of the same name.\n",
        " |      \n",
        " |      Optional argument chars defines which characters to remove.  If\n",
        " |      omitted or None (default) then as for the python string method,\n",
        " |      this defaults to removing any white space.\n",
        " |      \n",
        " |      e.g. print my_seq.strip(\"-\")\n",
        " |      \n",
        " |      See also the lstrip and rstrip methods.\n",
        " |  \n",
        " |  tomutable(self)\n",
        " |      Returns the full sequence as a MutableSeq object.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAAL\",\n",
        " |      ...              IUPAC.protein)\n",
        " |      >>> my_seq\n",
        " |      Seq('MKQHKAMIVALIVICITAVVAAL', IUPACProtein())\n",
        " |      >>> my_seq.tomutable()\n",
        " |      MutableSeq('MKQHKAMIVALIVICITAVVAAL', IUPACProtein())\n",
        " |      \n",
        " |      Note that the alphabet is preserved.\n",
        " |  \n",
        " |  tostring(self)\n",
        " |      Returns the full sequence as a python string (semi-obsolete).\n",
        " |      \n",
        " |      Although not formally deprecated, you are now encouraged to use\n",
        " |      str(my_seq) instead of my_seq.tostring().\n",
        " |  \n",
        " |  transcribe(self)\n",
        " |      Returns the RNA sequence from a DNA sequence. New Seq object.\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC\n",
        " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\",\n",
        " |      ...                  IUPAC.unambiguous_dna)\n",
        " |      >>> coding_dna\n",
        " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())\n",
        " |      >>> coding_dna.transcribe()\n",
        " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG', IUPACUnambiguousRNA())\n",
        " |      \n",
        " |      Trying to transcribe a protein or RNA sequence raises an exception:\n",
        " |      \n",
        " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
        " |      >>> my_protein.transcribe()\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      ValueError: Proteins cannot be transcribed!\n",
        " |  \n",
        " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False)\n",
        " |      Turns a nucleotide sequence into a protein sequence. New Seq object.\n",
        " |      \n",
        " |      This method will translate DNA or RNA sequences, and those with a\n",
        " |      nucleotide or generic alphabet.  Trying to translate a protein\n",
        " |      sequence raises an exception.\n",
        " |      \n",
        " |      Arguments:\n",
        " |       - table - Which codon table to use?  This can be either a name\n",
        " |                 (string), an NCBI identifier (integer), or a CodonTable\n",
        " |                 object (useful for non-standard genetic codes).  This\n",
        " |                 defaults to the \"Standard\" table.\n",
        " |       - stop_symbol - Single character string, what to use for terminators.\n",
        " |                       This defaults to the asterisk, \"*\".\n",
        " |       - to_stop - Boolean, defaults to False meaning do a full translation\n",
        " |                   continuing on past any stop codons (translated as the\n",
        " |                   specified stop_symbol).  If True, translation is\n",
        " |                   terminated at the first in frame stop codon (and the\n",
        " |                   stop_symbol is not appended to the returned protein\n",
        " |                   sequence).\n",
        " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
        " |               this checks the sequence starts with a valid alternative start\n",
        " |               codon (which will be translated as methionine, M), that the\n",
        " |               sequence length is a multiple of three, and that there is a\n",
        " |               single in frame stop codon at the end (this will be excluded\n",
        " |               from the protein sequence, regardless of the to_stop option).\n",
        " |               If these tests fail, an exception is raised.\n",
        " |      \n",
        " |      e.g. Using the standard table:\n",
        " |      \n",
        " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
        " |      >>> coding_dna.translate()\n",
        " |      Seq('VAIVMGR*KGAR*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
        " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
        " |      Seq('VAIVMGR@KGAR@', HasStopCodon(ExtendedIUPACProtein(), '@'))\n",
        " |      >>> coding_dna.translate(to_stop=True)\n",
        " |      Seq('VAIVMGR', ExtendedIUPACProtein())\n",
        " |      \n",
        " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
        " |      \n",
        " |      >>> coding_dna.translate(table=2)\n",
        " |      Seq('VAIVMGRWKGAR*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
        " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
        " |      Seq('VAIVMGRWKGAR', ExtendedIUPACProtein())\n",
        " |      \n",
        " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
        " |      this sequence could be a complete CDS:\n",
        " |      \n",
        " |      >>> coding_dna.translate(table=2, cds=True)\n",
        " |      Seq('MAIVMGRWKGAR', ExtendedIUPACProtein())\n",
        " |      \n",
        " |      It isn't a valid CDS under NCBI table 1, due to both the start codon and\n",
        " |      also the in frame stop codons:\n",
        " |      \n",
        " |      >>> coding_dna.translate(table=1, cds=True)\n",
        " |      Traceback (most recent call last):\n",
        " |          ...\n",
        " |      TranslationError: First codon 'GTG' is not a start codon\n",
        " |      \n",
        " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
        " |      has no effect:\n",
        " |      \n",
        " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
        " |      >>> coding_dna2.translate()\n",
        " |      Seq('LAIVMGR', ExtendedIUPACProtein())\n",
        " |      >>> coding_dna2.translate(to_stop=True)\n",
        " |      Seq('LAIVMGR', ExtendedIUPACProtein())\n",
        " |      \n",
        " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
        " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
        " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
        " |      \n",
        " |      NOTE - Does NOT support gapped sequences.\n",
        " |      \n",
        " |      NOTE - This does NOT behave like the python string's translate\n",
        " |      method.  For that use str(my_seq).translate(...) instead.\n",
        " |  \n",
        " |  ungap(self, gap=None)\n",
        " |      Return a copy of the sequence without the gap character(s).\n",
        " |      \n",
        " |      The gap character can be specified in two ways - either as an explicit\n",
        " |      argument, or via the sequence's alphabet. For example:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import generic_dna\n",
        " |      >>> my_dna = Seq(\"-ATA--TGAAAT-TTGAAAA\", generic_dna)\n",
        " |      >>> my_dna\n",
        " |      Seq('-ATA--TGAAAT-TTGAAAA', DNAAlphabet())\n",
        " |      >>> my_dna.ungap(\"-\")\n",
        " |      Seq('ATATGAAATTTGAAAA', DNAAlphabet())\n",
        " |      \n",
        " |      If the gap character is not given as an argument, it will be taken from\n",
        " |      the sequence's alphabet (if defined). Notice that the returned sequence's\n",
        " |      alphabet is adjusted since it no longer requires a gapped alphabet:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC, Gapped, HasStopCodon\n",
        " |      >>> my_pro = Seq(\"MVVLE=AD*\", HasStopCodon(Gapped(IUPAC.protein, \"=\")))\n",
        " |      >>> my_pro\n",
        " |      Seq('MVVLE=AD*', HasStopCodon(Gapped(IUPACProtein(), '='), '*'))\n",
        " |      >>> my_pro.ungap()\n",
        " |      Seq('MVVLEAD*', HasStopCodon(IUPACProtein(), '*'))\n",
        " |      \n",
        " |      Or, with a simpler gapped DNA example:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import IUPAC, Gapped\n",
        " |      >>> my_seq = Seq(\"CGGGTAG=AAAAAA\", Gapped(IUPAC.unambiguous_dna, \"=\"))\n",
        " |      >>> my_seq\n",
        " |      Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
        " |      >>> my_seq.ungap()\n",
        " |      Seq('CGGGTAGAAAAAA', IUPACUnambiguousDNA())\n",
        " |      \n",
        " |      As long as it is consistent with the alphabet, although it is redundant,\n",
        " |      you can still supply the gap character as an argument to this method:\n",
        " |      \n",
        " |      >>> my_seq\n",
        " |      Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
        " |      >>> my_seq.ungap(\"=\")\n",
        " |      Seq('CGGGTAGAAAAAA', IUPACUnambiguousDNA())\n",
        " |      \n",
        " |      However, if the gap character given as the argument disagrees with that\n",
        " |      declared in the alphabet, an exception is raised:\n",
        " |      \n",
        " |      >>> my_seq\n",
        " |      Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
        " |      >>> my_seq.ungap(\"-\")\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      ValueError: Gap '-' does not match '=' from alphabet\n",
        " |      \n",
        " |      Finally, if a gap character is not supplied, and the alphabet does not\n",
        " |      define one, an exception is raised:\n",
        " |      \n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> from Bio.Alphabet import generic_dna\n",
        " |      >>> my_dna = Seq(\"ATA--TGAAAT-TTGAAAA\", generic_dna)\n",
        " |      >>> my_dna\n",
        " |      Seq('ATA--TGAAAT-TTGAAAA', DNAAlphabet())\n",
        " |      >>> my_dna.ungap()\n",
        " |      Traceback (most recent call last):\n",
        " |         ...\n",
        " |      ValueError: Gap character not given and not defined in alphabet\n",
        " |  \n",
        " |  upper(self)\n",
        " |      Returns an upper case copy of the sequence.\n",
        " |      \n",
        " |      >>> from Bio.Alphabet import HasStopCodon, generic_protein\n",
        " |      >>> from Bio.Seq import Seq\n",
        " |      >>> my_seq = Seq(\"VHLTPeeK*\", HasStopCodon(generic_protein))\n",
        " |      >>> my_seq\n",
        " |      Seq('VHLTPeeK*', HasStopCodon(ProteinAlphabet(), '*'))\n",
        " |      >>> my_seq.lower()\n",
        " |      Seq('vhltpeek*', HasStopCodon(ProteinAlphabet(), '*'))\n",
        " |      >>> my_seq.upper()\n",
        " |      Seq('VHLTPEEK*', HasStopCodon(ProteinAlphabet(), '*'))\n",
        " |      \n",
        " |      This will adjust the alphabet if required. See also the lower method.\n",
        " |  \n",
        " |  ----------------------------------------------------------------------\n",
        " |  Data descriptors defined here:\n",
        " |  \n",
        " |  __dict__\n",
        " |      dictionary for instance variables (if defined)\n",
        " |  \n",
        " |  __weakref__\n",
        " |      list of weak references to the object (if defined)\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 31
    }
   ],
   "metadata": {}
  }
 ]
}